### ISSUE TITLE
trying to test if the expected result is not empty

### ISSUE THREAD

**Issue by hschmull** [xspec/xspec#230]
testing a docbook to xhtml transformation, l just wanted to make shure that my test is not comparing empty sequences, so I added:
```xml
<x:expect label="result must not be empty" test="node()"/>
```
test fails as expected with empty `x:context`, but looking into the generated result.xml I find the wole XSPEC file inside the `x:result`-Element:

```xml
<x:test successful="false">
         <x:label>result must not be empty</x:label>
         <x:result>
            <x:description stylesheet="file:/C:/Users/pagina/GitHub/xspec-test/xslt/docbook-to-html.xsl">
            <!-- ...-->
```
I expected the `x:expect/@select="node()"` to act as in XSLT with the result of my transformation as context node, so for instance returning an element `<p>` when succeding or an empty sequence (or `false()`) when failing.

### My setup:

Input-XML
```xml
<para xmlns="http://docbook.org/ns/docbook" xmlns:xlink="http://www.w3.org/1999/xlink" version="5.0">XSpec is a unit test and <link xlink:href="https://en.wikipedia.org/wiki/Behavior_Driven_Development">behaviour driven development</link> (BDD) framework for XSLT, XQuery and Schematron.</para>
```
XSLT
```xml
<xsl:stylesheet xmlns="http://www.w3.org/1999/xhtml"
	xmlns:xsl="http://www.w3.org/1999/XSL/Transform"
	xmlns:xlink="http://www.w3.org/1999/xlink"
	exclude-result-prefixes="#all"
	xpath-default-namespace="http://docbook.org/ns/docbook"
	version="3.0">
	
	<xsl:output method="xhtml" encoding="UTF-8" html-version="5.0"/>
	
	<xsl:template match="para">
		<p>
			<xsl:apply-templates/>
		</p>
	</xsl:template>
	
	<xsl:template match="link">
		<a href="{@xlink:href}">
			<xsl:apply-templates/>
		</a>
	</xsl:template>
	
</xsl:stylesheet>
```

XSPEC file
```xml
<x:description xmlns:x="http://www.jenitennison.com/xslt/xspec" 
    xmlns:xlink="http://www.w3.org/1999/xlink"
    xmlns:pa="https://www.pagina.gmbh"
    stylesheet="file:/C:/Users/pagina/GitHub/xspec-test/xslt/docbook-to-html.xsl"
    xmlns="http://docbook.org/ns/docbook">
    
    <x:scenario label="template with match 'para' literal">
        <x:context select="/para/link">
            <para>XSpec is a unit test and <link xlink:href="https://en.wikipedia.org/wiki/Behavior_Driven_Development">behaviour driven development</link> (BDD) framework for XSLT, XQuery and Schematron. It is based on the Spec framework of <link xlink:href="http://rspec.info/">RSpec</link>, which is a BDD framework for Ruby.</para>
        </x:context>
        
        <x:expect label="result must not be empty" test="node()"/>
    </x:scenario>
    
</x:description>
```

**Comment by AirQuick** [comment]
Actually the test must be Success, because `x:context` results in an empty sequence (because of its `@select` namespace) and `x:expect` expects no node (because of non-boolean `@test` and lack of `@select` or child node). I'll look into this.

**Comment by AirQuick** [comment]
By the way, maybe you expects that `x:expect/@test` works like `xsl:if/@test`. But actually it doesn't.

So, in your `x:expect`,
```xml
<x:expect label="result must not be empty" test="node()"/>
```
`@label` and `@test` contradict each other.

It should be either
```xml
<x:expect label="result must not be empty" test="exists($x:result)"/>
```
or
```xml
<x:expect label="result must have child node" test="exists(node())"/>
```
or
```xml
<x:expect label="result must not have child node" test="empty(node())"/>
```
whichever you intend.

(The last one can be
```xml
<x:expect label="result must not have child node" test="node()"/>
```
but not recommended.)

**Comment by AirQuick** [comment]
> looking into the generated result.xml I find the wole XSPEC file inside the x:result-Element

> Actually the test must be Success

Filed them as #234 with a simpler case and another case.

---
@hschmull
Despite #234, I think if you follow my [previous comment](https://github.com/xspec/xspec/issues/230#issuecomment-407430343), you can make `x:expect` work as you wish.

**Comment by hschmull** [comment]
Hello @AirQuick and thanks for your elaborate answer!
> maybe you expect that `x:expect/@test` works like `xsl:if/@test`.

That's true. In fact I'm a beginner in XSpec and just tryied in a way I was used to in XSLT. By failing and having your comment I learned a lot.
I understand I should use `exists()` but why do I have to refer to the result of my tested template by `$x:result` – shouldn't that be the context of `x:expect` by _default_ ?

**Comment by AirQuick** [comment]
The current behavior of `@test` has its own benefits. To demonstrate it,

This XSLT template generates `<A/><B/>Hello, World!`:

**test.xsl**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<xsl:stylesheet exclude-result-prefixes="#all" version="2.0"
    xmlns:xs="http://www.w3.org/2001/XMLSchema" xmlns:xsl="http://www.w3.org/1999/XSL/Transform">

    <xsl:output omit-xml-declaration="yes" />

    <xsl:template as="node()+" name="get-nodes">
        <A />
        <B />
        <xsl:text>Hello, </xsl:text>
        <xsl:text>World!</xsl:text>
    </xsl:template>

</xsl:stylesheet>
```

And this XSpec tests the template:

**test.xspec**
```xml
<?xml version="1.0" encoding="UTF-8"?>
<x:description stylesheet="test.xsl" xmlns:x="http://www.jenitennison.com/xslt/xspec">
    <x:scenario label="when result is a sequence of multiple nodes">
        <x:call template="get-nodes" />
        <x:expect label="$x:result holds the actual result as it is"
            test="
                ($x:result instance of node()+)
                and (count($x:result) eq 4)" />
        <x:expect label="@test context is a document node which wraps the actual result"
            test="
                (. instance of document-node())
                and exists(./A)
                and exists(./B)
                and exists(./text())" />

        <x:scenario label="the current behavior of @test makes these possible">
            <x:expect label="A can be referred by A" test="A"><A /></x:expect>
            <x:expect label="B can be found as a following-sibling of A"
                test="A/following-sibling::B"><B /></x:expect>
            <x:expect label="text nodes are merged" test="node()[last()]">Hello, World!</x:expect>
        </x:scenario>

        <x:scenario label="$x:result is not always intuitive">
            <x:expect label="A cannot be referred by A; it's self::A." test="$x:result/self::A"
                ><A /></x:expect>
            <x:expect
                label="B cannot be found as a following-sibling of A, because they don't share the common parent"
                test="empty($x:result/self::A/following-sibling::B)" />
            <x:expect label="text nodes are not merged" test="$x:result[last()]">World!</x:expect>
        </x:scenario>
    </x:scenario>
</x:description>
```

All the `x:expect` instances result in Success.

If you look into `x:expect` instances in the `the current behavior of @test makes these possible` scenario, you'll find them convenient or intuitive.

On the other hand, `x:expect` instances in the `$x:result is not always intuitive` scenario are somewhat awkward or counterintuitive.

**Comment by cirulls** [comment]
Thanks @hschmull and @AirQuick for this exchange, it may be useful to others too. 

@hschmull: apart from the limitations highlighted in #234, does @AirQuick 's reply address your issue? Can this issue be closed or is addressed by #234? Thanks!

**Comment by AirQuick** [comment]
Closing this. Please feel free to open another issue if you have any.
